packages/shared/src/features/onboarding/hooks/useFunnelNavigation.ts

```diff
@@ -0,0 +1,177 @@
+import { useMemo, useState, useEffect, useCallback } from 'react';
+import { useSearchParams, useRouter, usePathname } from 'next/navigation';
+import { useAtom } from 'jotai/react';
+import { UNDO } from 'jotai-history';
+import type {
+  FunnelJSON,
+  FunnelPosition,
+  NonChapterStep,
+} from '../types/funnel';
+import { FunnelStepTransitionType } from '../types/funnel';
+import type { TrackOnNavigate } from './useFunnelTracking';
+import {
+  funnelPositionAtom,
+  getFunnelStepByPosition,
+  funnelPositionHistoryAtom,
+} from '../store/funnelStore';
+
+interface UseFunnelNavigationProps {
+  funnel: FunnelJSON;
+  onNavigation: TrackOnNavigate;
+}
+
+type Chapters = Array<{ steps: number }>;
+type StepMap = Record<NonChapterStep['id'], { position: FunnelPosition }>;
+type NavigateFunction = (options: {
+  to: NonChapterStep['id'];
+  type?: FunnelStepTransitionType;
+}) => void;
+
+interface HeaderNavigation {
+  hasTarget: boolean;
+  navigate: () => void;
+}
+
+export interface UseFunnelNavigationReturn {
+  chapters: Chapters;
+  navigate: NavigateFunction;
+  position: FunnelPosition;
+  step: NonChapterStep;
+  back: HeaderNavigation;
+  skip: HeaderNavigation;
+}
+
+export const useFunnelNavigation = ({
+  funnel,
+  onNavigation,
+}: UseFunnelNavigationProps): UseFunnelNavigationReturn => {
+  const router = useRouter();
+  const searchParams = useSearchParams();
+  const pathname = usePathname();
+  const [stepTimerStart, setStepTimerStart] = useState<number>(0);
+  const [position, setPosition] = useAtom(funnelPositionAtom);
```

Comment:
I guess we need atoms to reshare between things here?
Don't we already have exposes context for funnels specifically?